### ISSUE TITLE
[Proposal] Blade Extension Classes

### ISSUE THREAD

**Issue by paulredmond** [laravel/internals#682]
I would like to contribute a simple blade extension system to laravel/framework which allows a Blade extension class to be defined and registered in the container.

Through tagging the Laravel framework could register all the blade directives and conditionals automatically. That means that registered providers could tag an extension and Laravel would wire it up in blade.

Here's a basic example of what a TwigExtension class would look like, allowing you to group related extensions together:

```php
<?php

/**
 * Example blade extension class for grouping extensions
 */
class ExampleBladeExtension implements BladeExtension
{
    public function getDirectives()
    {
        return [
            'truncate' => [$this, 'truncate']
        ];
    }

    public function getConditionals()
    {
        return [
            'prod' => [$this, 'isProduction']
        ];
    }

    public function isProduction()
    {
        return app()->environment('production');
    }

    /**
     * @see https://zaengle.com/blog/exploring-laravels-custom-blade-directives
     */
    public function truncate($expression)
    {
        list($string, $length) = explode(',',str_replace(['(',')',' '], '', $expression));

        return "<?php echo e(strlen({$string}) > {$length} ? substr({$string},0,{$length}).'...' : {$string}); ?>";
    }
}
```

Here are a couple resources for what I mean:

- [Basic idea](https://gist.github.com/paulredmond/8ba3c45e0374dc1fd775fe91b92b93e9) of what I mean with pseudo code

- Making a [Twig Extension in Symfony](https://symfony.com/doc/current/templating/twig_extension.html)

- [Example](https://twitter.com/paulredmond/status/885261938023776257) where I register the directives in a Service provider `boot()` method. This would come from the framework during boot, I just experimented in an `AppServiceProvider` as an example:

```php
collect($this->app->tagged('blade.extension'))->each(function ($extension) {
    collect($extension->getDirectives())->each(function ($callable, $name) {
        Blade::directive($name, $callable);
    });

    collect($extension->getConditionals())->each(function ($callable, $name) {
        Blade::if($name, $callable);
    });
});
```

Example extension getting registered in as service provider's `register()` method:

```php
// Service container available to extensions
$this->app->bind(ExampleExtension::class, function ($app) {
    return new ExampleExtension();
});

$this->app->tag([ExampleExtension::class], 'blade.extension');
```

I have this working in a test Laravel 5.5 app, and I would love to contribute the PR for this if accepted. I also realize that not everyone might want to have/use this feature, let me know what you think!

**Comment by m1guelpf** [comment]
This is awesome!

**Comment by paulredmond** [comment]
I've added a little bit of code to my fork if you want to see where I am headed with it and if this is something that would be useful to the framework 👍 

https://github.com/laravel/framework/compare/master...paulredmond:master?expand=1

**Comment by paulredmond** [comment]
@taylorotwell and @themsaid I was hoping to get a little feedback from you guys on this idea. I added a little code to my fork to demonstrate, I appreciate your time.

As I noted above, I am not offended if this isn't desired in framework, just an idea I had that I used in my own application.

Comparing my fork against master gets you the diff - https://github.com/laravel/framework/compare/master...paulredmond:master?expand=1
